hmpps-sqs-spring-boot-autoconfigure/src/main/kotlin/uk/gov/justice/hmpps/sqs/telemetry/TraceInjectingExecutionInterceptor.kt

```diff
@@ -0,0 +1,75 @@
+package uk.gov.justice.hmpps.sqs.telemetry
+
+import io.opentelemetry.api.GlobalOpenTelemetry
+import io.opentelemetry.api.trace.Span
+import io.opentelemetry.api.trace.SpanKind
+import io.opentelemetry.instrumentation.annotations.WithSpan
+import software.amazon.awssdk.core.interceptor.Context
+import software.amazon.awssdk.core.interceptor.ExecutionAttributes
+import software.amazon.awssdk.core.interceptor.ExecutionInterceptor
+import software.amazon.awssdk.services.sns.model.PublishBatchRequest
+import software.amazon.awssdk.services.sns.model.PublishRequest
+import software.amazon.awssdk.services.sqs.model.SendMessageBatchRequest
+import software.amazon.awssdk.services.sqs.model.SendMessageRequest
+import io.opentelemetry.context.Context as SpanContext
+import software.amazon.awssdk.services.sns.model.MessageAttributeValue as SnsMessageAttributeValue
+import software.amazon.awssdk.services.sqs.model.MessageAttributeValue as SqsMessageAttributeValue
+
+/**
+ * Intercepts `PublishRequest` and `SendMessageRequest` requests, and
+ *  1. creates an OpenTelemetry span with the name "PUBLISH $eventType",
+ *  2. injects the trace context into the message attributes before sending.
+ */
+class TraceInjectingExecutionInterceptor : ExecutionInterceptor {
+  override fun modifyRequest(context: Context.ModifyRequest?, executionAttributes: ExecutionAttributes?) =
+    when (val request = context?.request()) {
+      is PublishRequest -> publishSpan(request.messageAttributes()["eventType"]?.stringValue()) {
+        request.toBuilder().messageAttributes(request.messageAttributes().withSnsTelemetryContext()).build()
+      }
+
+      is PublishBatchRequest -> publishSpan {
+        request.toBuilder().publishBatchRequestEntries(
+          request.publishBatchRequestEntries().map { entry ->
+            entry.toBuilder().messageAttributes(entry.messageAttributes().withSnsTelemetryContext()).build()
+          },
+        ).build()
+      }
+
+      is SendMessageRequest -> publishSpan(request.messageAttributes()["eventType"]?.stringValue()) {
+        request.toBuilder().messageAttributes(request.messageAttributes().withSqsTelemetryContext()).build()
+      }
+
+      is SendMessageBatchRequest -> publishSpan {
+        request.toBuilder().entries(
+          request.entries().map { entry ->
+            entry.toBuilder().messageAttributes(entry.messageAttributes().withSqsTelemetryContext()).build()
+          },
+        ).build()
+      }
+
+      else -> request
+    }
+
+  /**
+   * Creates a "PUBLISH" span and injects
+   */
+  @WithSpan(kind = SpanKind.PRODUCER)
+  private fun <T> publishSpan(eventType: String? = null, modifiedRequest: () -> T): T {
+    Span.current().updateName(eventType?.let { "PUBLISH $it" } ?: "PUBLISH")
+    return modifiedRequest()
+  }
```

Comment:
We've been using the @WithSpan annotation on our listeners to get message receivers into App Insights, e.g. [like this](https://github.com/ministryofjustice/hmpps-prisoner-from-nomis-migration/blob/main/src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerfromnomismigration/activities/ActivitiesMigrationMessageListener.kt#L29).

I guess this means we'll be able to drop that annotation now?